### ISSUE TITLE
Regression: unification error in function application

### ISSUE THREAD

**Issue by sstucki** [agda/agda#3148]
This bug affects Agda >= 2.5.4.

I encountered this issue when trying to upgrade a project from Agda 2.5.3 to 2.5.4. The code type checks under 2.5.3 but in 2.5.4 I get a strange type-inference/unification error. I spent 2h trying to minimize the program but did not succeed, so I'm reporting it here as found in the original project, which is available [on GitHub](https://github.com/sstucki/f-omega-int-agda/tree/master/src).

The error occurs in the module `FOmegaInt.Kinding.Simple.Normalization` [on line 76](https://github.com/sstucki/f-omega-int-agda/blob/master/src/FOmegaInt/Kinding/Simple/Normalization.agda#L76):

```
(nfCtx (_Γ_2090 k-kd a∈*)) != (_2098 k-kd a∈* E.∷ _2099 k-kd a∈*)
of type
(.Data.Fin.Substitution.Context.CtxExt (KdOrTp Elim) 0 (suc .n))
when checking that the inferred type of an application
  ⌊ nfCtx (_Γ_2090 k-kd a∈*) ⌋Ctx ⊢Nf
  nf (nfCtx (_Γ_2090 k-kd a∈*)) (_a_2091 k-kd a∈*) ∈
  ⌊ _k_2092 k-kd a∈* ⌋
matches the expected type
  kd ⌊ nfKind (nfCtx .Γ) .k ⌋ E.∷ ⌊ nfCtx .Γ ⌋Ctx ⊢Nf
  nf (kd (nfKind (nfCtx .Γ) .k) ElimCtx.∷ nfCtx .Γ) .a ∈ ★
```

From what I can tell, the offending term is `nf-Tp∈ a∈*`, where the type checker somehow fails to unify the implicit argument `Γ` of `nf-Tp∈` with the first index `kd .k ∷ .Γ` of the argument `a∈*`. If it managed to do so, the inferred type of the application would be `⌊ kd (nfKind (nfCtx .Γ) .k) E.∷ nfCtx .Γ ⌋Ctx ⊢Nf nf (kd (nfKind (nfCtx .Γ) .k) E.∷ nfCtx .Γ) .a ∈ ★` as required.

Interestingly, I can avoid the type error by locally `let`-binding the application and immediately returning it, like so:

```Agda
... = ∈-∀-f (nf-kd k-kd) (let dummy = nf-Tp∈ a∈* in dummy)
```
A similar error also occurs on lines 81 and 94 of the same file.

**Comment by asr** [comment]
I don't know the reason of the regression, but I suggest to create a branch with the changes required so far (e.g. to add the signatures `ℕ → Set`) for reproducing the issue.

**Comment by sstucki** [comment]
@asr, here you go: https://github.com/sstucki/f-omega-int-agda/tree/agda-issue-3148

As per your suggestion, that branch just contains the changes necessary to make it compatible with stdlib 0.16.

And here's one where I applied the workaround described above (insert dummy `let`-bindings): https://github.com/sstucki/f-omega-int-agda/tree/wip/upgrade-2.5.4

**Comment by asr** [comment]
Note that using Agda 2.5.4 and the `agda-issue-3148` branch I got the error:

```
$ agda -i src src/README.agda
...
<some-path>/f-omega-int-agda/src/FOmegaInt/Syntax.agda:12,8-48
Failed to solve the following constraints:
  .Agda.Primitive.lzero = _ℓ_275
  _ℓ_275 = .Agda.Primitive.lzero
  .Agda.Primitive.lsuc _ℓ_275
    = .Agda.Primitive.lsuc .Agda.Primitive.lzero
when scope checking the declaration
  import Data.Fin.Substitution.Context as Context
<some-path>/f-omega-int-agda/src/FOmegaInt/Syntax.agda:12,8-48
Unsolved metas at the following locations:
  <some-path>/f-omega-int-agda/src/Data/Fin/Substitution/Context.agda:155,35-44
when scope checking the declaration
  import Data.Fin.Substitution.Context as Context
```

**Comment by sstucki** [comment]
@asr: whoops, sorry about that.

I had only tested the branch using Agda 2.6.0, but apparently Agda 2.5.4 requires some extra annotations. I updated the two branches, so they should work with both Agda 2.5.4 and 2.6.0 (and stdlib 0.16) now.

(Note that your local versions of these branches will have diverged as I rebased and force-pushed the changes.)

**Comment by asr** [comment]
Thanks! As I said, I don't know the reason of the regression, but I *easily* could reproduce the issue with both Agda 2.5.4 and Agda 2.6.0-0ad14b8 using the `agda-issue-3148` branch:

```
$ agda -i src src/README.agda
...
<some-path>/f-omega-int-agda/src/FOmegaInt/Kinding/Simple/Normalization.agda:76,49-59
nfCtx (_Γ_113 k-kd a∈*) != _121 k-kd a∈* E.∷ _122 k-kd a∈* of type
.Data.Fin.Substitution.Context.CtxExt (KdOrTp Elim) 0 (suc .n)
when checking that the inferred type of an application
  ⌊ nfCtx (_Γ_113 k-kd a∈*) ⌋Ctx ⊢Nf
  nf (nfCtx (_Γ_113 k-kd a∈*)) (_a_114 k-kd a∈*) ∈
  ⌊ _k_115 k-kd a∈* ⌋
matches the expected type
  kd ⌊ nfKind (nfCtx .Γ) .k ⌋ E.∷ ⌊ nfCtx .Γ ⌋Ctx ⊢Nf
  nf (kd (nfKind (nfCtx .Γ) .k) ElimCtx.∷ nfCtx .Γ) .a ∈ ★

**Comment by asr** [comment]
> I had only tested the branch using Agda 2.6.0, but apparently Agda 2.5.4 requires some extra annotations.

I'm not aware of this situation. To avoid a possible future regression or to improve the documentation, could you create a new issue reporting this behaviour, please.

**Comment by sstucki** [comment]
> To avoid a possible future regression or to improve the documentation, could you create a new issue reporting this behaviour, please.

Sure, I can do that. Do you want me to find a minimal example or just create two more branches of my project on Github and report them in a new issue?

To be clear: this is just about 2.6.0 being able to infer a few more type signatures automaticall than 2.5.4, so it's not strictly speaking a bug.

**Comment by asr** [comment]
I MWE would be great.

**Comment by sstucki** [comment]
> I MWE would be great.

Reported in #3153.

**Comment by jespercockx** [comment]
Without a smaller example, I'm afraid this won't be fixed for 2.6.0.

**Comment by sstucki** [comment]
I can no longer reproduce this bug in Agda 2.6.0.1 (with stdlib 1.0.1) so I suggest marking it as resolved.
